### PR TITLE
Fix filter functionality of Worksheets after sort/pagination

### PR DESCRIPTION
**Pull request by ramonski**
## Description of the issue/feature this PR addresses

Linked issue: https://github.com/senaite/senaite.core/issues/699

The primary goal of this PR was actually only to fix the filter fields of the worksheet /add_analyses view...

This PR refactors now the whole javascript logic located in `bika.lims.worksheet.js` in the corresponding `bika.lims.worksheet.coffee`.

Therefore, the following functionality should be given (categorized by view -> JS controller class):

###  `/worksheets` -> `bika.lims.worksheet.WorksheetFolderView`

- [x] Selecting a template automatically selects the assigned instrument of the template in the selection
- [x] Selecting an instrument displays the notification "Only the analyses for which the selected instrument is allowed will be added automatically."

### `/add_analyses` -> `bika.lims.worksheet.WorksheetAddAnalysesView`

- [x] Selecting a service category filters the selection of the services accordingly
- [x] Clicking the search button filters accordingly
- [x] Clicking the table sorters or the "show more" button persists the filters

### `/add_control` or `/add_blank` -> `bika.lims.worksheet.WorksheetAddQCAnalysesView`

- [x] Selecting a service by clicking on the checkbox loads the reference sample listing table
- [x] Clicking on a row of the reference sample listing table adds the reference sample to the worksheet

### `/add_duplicate` -> `bika.lims.worksheet.WorksheetAddDuplicateAnalysesView`

- [x] Clicking on a row adds the duplicate

### `/manage_results` -> `bika.lims.worksheet.WorksheetManageResultsView`

- [ ] Automatically loads the instruments and methods constraints (really weird logic, I didn't understand these flags and what happens actually)
- [x] Changing the Analyst automatically saves to server and gives a notification message
- [x] Changing the Instrument in the header saves it to the server and give as notification message
- [ ] Changing the Method of an Analysis loads analysis and method constraints
- [ ] Changing the Instrument of an Analysis disables the instrument for other analyses, enables the previous instrument elsewhere and gives the None option
- [x] Changing the detection limit of an Analysis controls somehow the behavior of the result and uncertainty fields, but I couldn't figure out the logic
- [x] Clicking on the remark balloon toggles the remark field
- [x] Changing the category selection of the wide-interims in the header, filters the interim fields which belong to this category
- [x] Changing the interim selection of the wide-interims in the header, set the value of the field
- [x] Clicking on the "Apply" button of the wide-interims saves according to the empty/non-empty checkbox

## Current behavior before PR

Filter settings (category, service, client) were lost after sorting or clicking the "show more" button in worksheet `/add_analyses` view.

## Desired behavior after PR is merged

Filter settings (category, service, client) persist after sorting or clicking the "show more" button in worksheet `/add_analyses` view.

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/browser/worksheet/templates/add_analyses.pt [hunk 1]
```diff
@@ -1,133 +1,137 @@
...
-	<fieldset/>
+          <!-- Search Button -->
+          <a tal:attributes="href view/view_url"
+             class="ws-analyses-search-button"
+             i18n:translate="">Search</a>
 
-	<div id="folderlisting-main-table"
-		tal:content="structure view/contents_table"/>
+        </fieldset>
+      </tal:searchform>
```
Comment (xispa) [inline bika/lims/browser/worksheet/templates/add_analyses.pt]: Before this PR, the filter elements were appearing in-line, now appear stacked. Probably because you wrapped the filter elements inside `<div class="field">..</div>` 's. 

I think it would be much better if we could keep the "traditional "inline" layout while assuring it gets rendered correctly when `senaite.lims` UI is installed. I suggest to include all elements inside a `<div class='row'>` and wrap each of them by a `<div class='col-sm-4 field'>` and add a css rule in `ploneCustom.css` similar to:
```css
.template-add_analyses.portaltype-worksheet .criteria .field {
    clear: none;
    float: left;
    padding-right: 20px;
}
```

File: bika/lims/browser/js/coffee/bika.lims.worksheet.coffee [hunk 2]
```diff
@@ -0,0 +1,1070 @@
...
+      $(resfld).val ""
+      $(resfld).prop "readonly", no
+      # Activate uncertainty?
+      if uncfld.length > 0
+        $(uncfld).val ""
+        $(uncfld).prop "readonly", no
+        $(uncfld).closest("td").children().show()
+
+    # Maybe the result is used in calculations...
+    $(resfld).change()
```
Comment (xispa) [inline bika/lims/browser/js/coffee/bika.lims.worksheet.coffee]: Indeed, this js takes into account two aspects about detection limits: the possibility to set a DL manually + relationship between DL and Calculations.

First, Manual entry of Detection Limits, see  https://github.com/bikalims/bika.lims/pull/1496

When "Manual Entry" is set to `True` the user can choose any of the following options from the selection list: `<`, `>` or empty. If the user selects empty, all works as usual: the result entered is the exact result obtained e.g. from the instrument or from the analytical technique used. Indeed, could happen this result is below the detection limit. No problem, `< ldl` will be displayed everywhere (including reports), hiding the submitted-"real" result for that analysis.

On the other hand, if the user selects `<` or `>`, the value he/she introduces is not the exact result obtained, rather he is telling something like "I obtained a result that I don't care about because is below *this* detection limit". Note the difference, in this case, the user does not set a "real" result, rather the detection limit itself. From a technical side, the only difference is that in the latter, the `Result` field from `Analysis`, has a "<" or ">" prepended to the value (not result, rather detection limit) submitted by the user.

Second, Detection Limits + Calculations, see https://github.com/bikalims/bika.lims/pull/1500

.Basically, detection limits can be used in calculations. Consider a formula like:
```python
"Below LDL!" if [AS2.BELOWLDL] else "Not below LDL!"
```
where `AS2` is the keyword of a given Analysis Service.

In summary, if the user changes the value of the DL selection list, we always need to notify that the result has changed, because other Analyses may need to update their result accordingly!.

File: bika/lims/browser/js/coffee/bika.lims.worksheet.coffee [hunk 3]
```diff
@@ -0,0 +1,1070 @@
...
+      $(resfld).val defdls["max"]
+      # Inactivate uncertainty?
+      if uncfld.length > 0
+        $(uncfld).val ""
+        $(uncfld).prop "readonly", yes
+        $(uncfld).closest("td").children().hide()
+    else
+      $(resfld).val ""
+      $(resfld).prop "readonly", no
+      # Activate uncertainty?
```
Comment (xispa) [inline bika/lims/browser/js/coffee/bika.lims.worksheet.coffee]: Yes, uncertainty must be only active if the result is a "real" result, not a detection limit (it does not make any sense to calculate the uncertainty if the result introduced is a detection limit instead of a discrete result!) - see this comment https://github.com/senaite/senaite.core/pull/721/files#r174639383 for further details

File: bika/lims/browser/js/coffee/bika.lims.worksheet.coffee [hunk 4]
```diff
@@ -0,0 +1,1070 @@
...
+    else if method_constraints[5] == 1
+      $(".interim input[uid='#{analysis_uid}']").prop "disabled", no
+      $("input[field='Result'][uid='#{analysis_uid}']").prop "disabled", no
+
+    # Info/Warn message?
+    $(".alert-instruments-invalid[uid='#{analysis_uid}']").remove()
+    if method_constraints[6] and method_constraints[6] != ""
+      $(i_selector).after "<img uid='#{analysis_uid}' class='alert-instruments-invalid' src='#{@get_portal_url()}/++resource++bika.lims.images/warning.png' title='#{method_constraints[6]}'>"
+    $(".amconstr[uid='#{analysis_uid}']").remove()
+
```
Comment (xispa) [inline bika/lims/browser/js/coffee/bika.lims.worksheet.coffee]: Indeed, this is the weirdest function of all them. Just a side note, all these "constraints" can be more understandable (or maybe is more confusing, though) if you look to this: https://github.com/senaite/senaite.core/blob/master/bika/lims/tests/halt/test_method_instrument_constraints.py#L116-L184
I know, is hell, but together with https://cloud.githubusercontent.com/assets/832627/15012945/c4bca092-11fc-11e6-90a3-4628e4fa7d08.png it was a good exercise to squeeze the brain a bit and maybe train it for complex sudoku.

File: bika/lims/browser/worksheet/ajax.py [hunk 5]
```diff
@@ -29,11 +29,18 @@ def __init__(self, context, request):
...
+            "portal_type": 'AnalysisService',
+            "inactive_state": 'active',
+            "sort_on": 'sortable_title',
+        }
+
+        getCategoryUID = self.request.get('getCategoryUID', '')
+        if getCategoryUID:
+            query["getCategoryUID"] = getCategoryUID
+
+        brains = bsc(query)
```
Comment (xispa) [inline bika/lims/browser/worksheet/ajax.py]: Remember `brains = api.search(query, "bika_setup_catalog")`:P

File: bika/lims/browser/worksheet/templates/add_analyses.pt [hunk 6]
```diff
@@ -1,133 +1,137 @@
...
+      <!-- Listing Table -->
+      <div id="folderlisting-main-table"
+           tal:content="structure view/contents_table"/>
 
-</metal:content-core>
+      <tal:remarks define="field python:context.Schema()['Remarks'];
+                           errors python:{};">
+        <p style="margin-top:2em;"/>
+        <metal:widget use-macro="python:context.widget('Remarks', mode='edit')" />
+      </tal:remarks>
```
Comment (xispa) [inline bika/lims/browser/worksheet/templates/add_analyses.pt]: I think it would be better if we remove this remarks field from this "Add Analyses" view.

**Comment by xispa** [comment]
Ummm, seems the method-instrument thing is not working properly. Give me some time to figure out what's going on (maybe is not related with your coffee).

File: bika/lims/browser/worksheet/templates/add_analyses.pt (hunk 1 above)
Comment (ramonski) [reply bika/lims/browser/worksheet/templates/add_analyses.pt]: perfect, will change that.

Since `ploneCustom.css` is not active in `senaite.lims`: https://github.com/senaite/senaite.lims/blob/master/src/senaite/lims/profiles/default/cssregistry.xml#L66

we can use in general all the bootstrap classes and divs and change it for the core just in `ploneCustom.css`, that's cool!

**Review by xispa — APPROVED** [review]
